### ISSUE TITLE
Bug: Using sentry_debug on a non-main thread before init may break Hub::main

### ISSUE THREAD

**Issue by Mcat12** [getsentry/sentry-rust#237]
Referenced code: [`hub.rs`]

**TL;DR**: Using `sentry_debug` may break `Hub::main` if it happens too early on a non-main thread. Using the `debug-log` feature may avoid this behavior.

[`Hub::main`] reads from a lazy-static [`PROCESS_HUB`] value. When `PROCESS_HUB` is first used, it is initialized with the thread ID of the current thread. It is assumed that the first usage of `PROCESS_HUB` is in the main function (via `sentry::init`), but this is not always the case.

There are also two thread-local values, [`THREAD_HUB`] and [`USE_PROCESS_HUB`]. `USE_PROCESS_HUB` holds a boolean which decides if the current thread should use the main thread's hub, aka `PROCESS_HUB`. These thread-locals are used by [`Hub::with`] to figure out if the hub should be taken from `PROCESS_HUB` or `THREAD_HUB`.

The issue occurs if the first usage of `PROCESS_HUB` is outside of the main thread. This initializes its thread ID to the current thread ID, which is not the main thread's ID. Future calls to `Hub::with` on the main thread will use `THREAD_HUB` instead of `PROCESS_HUB`. Since the Sentry client is initialized on the main thread, and the main thread is not using `PROCESS_HUB`, the initialized client is inaccessible to other threads.

The concrete way that this occurs in my code is when I provide a non-default transport in the client options (Edit: happens with the default reqwest transport as well):
```rust
let curl_transport_factory = |options: &sentry::ClientOptions| {
    Arc::new(sentry::transports::CurlHttpTransport::new(options)) as Arc<dyn sentry::Transport>
};
let options = sentry::ClientOptions {
    transport: Some(Arc::new(curl_transport_factory)),
    release: sentry::release_name!(),
    ..sentry::ClientOptions::default()
};
let client = sentry::init(options);
```

[The transport factory method is called before the client is bound to the hub][create_transport]. [`CurlHttpTransport::new`] spawns a thread which immediately calls [`sentry_debug`]. `sentry_debug` uses `log::debug` when the `debug-logs` feature is enabled (a non-default feature), but otherwise uses `Hub::with`. As long as the new thread reaches the `Hub::with` call before the main thread accesses `PROCESS_HUB`, this bug is triggered.

[`hub.rs`]: https://github.com/getsentry/sentry-rust/blob/36218ea5e815d871c073ee15d7f71ee01dab69c7/sentry-core/src/hub.rs
[`Hub::main`]: https://github.com/getsentry/sentry-rust/blob/36218ea5e815d871c073ee15d7f71ee01dab69c7/sentry-core/src/hub.rs#L132
[`Hub::with`]: https://github.com/getsentry/sentry-rust/blob/36218ea5e815d871c073ee15d7f71ee01dab69c7/sentry-core/src/hub.rs#L141
[`PROCESS_HUB`]: https://github.com/getsentry/sentry-rust/blob/36218ea5e815d871c073ee15d7f71ee01dab69c7/sentry-core/src/hub.rs#L19
[`THREAD_HUB`]: https://github.com/getsentry/sentry-rust/blob/36218ea5e815d871c073ee15d7f71ee01dab69c7/sentry-core/src/hub.rs#L27
[`USE_PROCESS_HUB`]: https://github.com/getsentry/sentry-rust/blob/36218ea5e815d871c073ee15d7f71ee01dab69c7/sentry-core/src/hub.rs#L29
[create_transport]: https://github.com/getsentry/sentry-rust/blob/36218ea5e815d871c073ee15d7f71ee01dab69c7/sentry-core/src/client.rs#L102
[`CurlHttpTransport::new`]: https://github.com/getsentry/sentry-rust/blob/36218ea5e815d871c073ee15d7f71ee01dab69c7/sentry/src/transport.rs#L289
[`sentry_debug`]: https://github.com/getsentry/sentry-rust/blob/36218ea5e815d871c073ee15d7f71ee01dab69c7/sentry-core/src/macros.rs#L56

**Comment by Swatinem** [comment]
Thanks for the very detailed report!

I actually tried reworking all of this early on in my `0.19` work because it is super hard to follow.

For this specific usecase, I think it would be enough to just do an empty `Hub::with` call before initializing the transport, right? That way the `PROCESS_HUB` at least exists, even though still empty.

**Comment by Mcat12** [comment]
Calling `Hub::with` manually makes sense, that should work. I chose to work around it by enabling `debug-logs` since I'm already using `slog`.
